FIR DFA: some slightly more questionable changes

pyos opened the pull request:
I had more all along.

The reason I'm submitting this as a separate PR is that (except for the first commit) is that I'm not entirely sure if *all* of this is a good idea. Some of the commits add new features that K1 did not have, and some make an architectural change to how PersistentFlow is built.

This is about all I have for now that is 100% working. After that I tried to fix the handling of postponed anonymous functions (again), but that turned out to be *really hard* (again).

^KT-22996 tag fixed-in-k2
^KT-7676 tag fixed-in-k2

pyos commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt:
```diff
@@ -1348,31 +1348,32 @@ class ControlFlowGraphBuilder {
+            //                            /-----------v
+            //   a ----> a.b ----> a?.b ----> c ----> a?.b ?: c
+            //       \------------------------^
+            // Only do this for control flow though - while technically data flow is the same,
+            // fewer implications are needed to implement the general version.
```
This is actually not true - if both data and control flow are directed to elvis RHS, then the safe call exit node has only one input (non-null receiver case) and thus needs no new implications *at all*. So the total number of implications in the safe call is the same, *and* the liveness of nodes is more correct. I'll do that separately together with fixing other cases where liveness is incorrect.

pyos replied:
Amended the commit since I had to rebase anyway.

demiurg906 commented:
@pyos could you please stop pushing to this PR? I started to review it and force pushes interrupt it

pyos commented:
Yeah, sorry

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/FirBuilderInferenceSession.kt:
```diff
@@ -269,9 +268,6 @@ class FirStubTypeTransformer(
         substitutor.substituteOrNull(resolvedTypeRef.type)?.let {
             resolvedTypeRef.withReplacedConeType(it)
         } ?: resolvedTypeRef
-
```
Can you please elaborate this change? This transformer has nothing to do with regular body resolve transformers and needed only for builder inference

pyos replied:
That change has no effect, the override does the same thing as `transformElement`. I removed it so that it doesn't appear in uses of `transformArguments`, which is what I was looking for.

demiurg906 replied:
Ok. It's worth to extract this change to separate commit

pyos replied:
done

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/InvocationKindTransformer.kt:
```diff
@@ -6,98 +6,49 @@
+        val lambda = argument.unwrapAnonymousFunctionExpression() ?: continue
+        val kind = byParameter[parameter] ?: EventOccurrencesRange.UNKNOWN.takeIf {
+            // Inline functional parameters have to be called in-place; that's the only permitted operation on them.
+            isInline && !parameter.isNoinline && !parameter.isCrossinline &&
```
Actually you should check only for `noinline` modifier here. `crossinline` keeps the lambda inline (so you are required to call it in place). The only difference is that `crossinline` prohibit non-local return from lambda body

demiurg906 replied:
Wow, TIL something new. Thank you

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt:
```diff
@@ -1115,41 +1108,34 @@ open class FirExpressionsResolveTransformer(transformer: FirAbstractBodyResolveT
-                delegatedConstructorCall.transformChildren(transformer, ResolutionMode.ContextDependent)
-            }
+        dataFlowAnalyzer.enterDelegatedConstructorCall()
+        val lastDispatchReceiver = implicitReceiverStack.lastDispatchReceiver()
```
Originally this `try/finally` was needed not for cases when there is an expression during resolution, but just for removal of need to call `dataFlowAnalyzer.exitDelegatedConstructorCall` on each return in this function (there were several returns some time ago)

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt:
```diff
@@ -112,56 +89,38 @@ class ControlFlowGraphBuilder {
+            // * UncaughtExceptionPath: last expression = whatever threw, *not* a return value
+            // * ReturnPath(this lambda): these go from `finally` blocks, so that's not the return value;
+            //   look in `nonDirectJumps` instead.
+            if (edge.kind.usedInCfa && edge.label == NormalPath) it.extractArgument() else null
```
`runIf(edge.kind.usedInCfa && edge.label == NormalPath) { it.extractArgument() }`

demiurg906 reviewed (changes requested):
Thank you for another great job!

pyos commented on compiler/tests-spec/testData/diagnostics/notLinked/dfa/neg/6.fir.kt:
```diff
@@ -93,9 +93,9 @@ fun case_10(x: Any?) {
-            <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>
-            <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>?.<!UNRESOLVED_REFERENCE!>prop_1<!>
-            <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>?.<!UNRESOLVED_REFERENCE!>prop_2<!>
+            <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any? & kotlin.Any")!>x<!>
```
If `x` is null, then `is SealedMixedChildObject1?` is true., so in the `else` branch `x` is, in fact, `Any`. `foo` should not be called in this code.

pyos replied:
I suppose this is somewhat confusing if you write code like

```kt
fun foo(x: String?) {
    if (x is Int?) return
    x.length // technically correct...
}
```

- `is Int?` is equivalent to `is Int? & String?`, meaning `is Nothing?`. Maybe this should be an IDE inspection - similar to "instance check always false", but more like "instance check can only succeed for null".

demiurg906 replied:
I rechecked and yes, you are right.

demiurg906 replied:
You are right, thanks

pyos commented on compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/Flow.kt:
```diff
@@ -5,8 +5,8 @@
-    abstract val knownVariables: Set<RealVariable>
-    abstract fun unwrapVariable(variable: RealVariable): RealVariable
-    abstract fun getTypeStatement(variable: RealVariable): TypeStatement?
+interface Flow {
```
Ok, this was an accidental change anyway I think - at some point I wanted to rename PersistentFlow to Flow, but then manually reverted that and forgot that Flow was an abstract class rather than an interface.

pyos commented on compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/LogicSystem.kt:
```diff
@@ -7,37 +7,32 @@ package org.jetbrains.kotlin.fir.resolve.dfa
-    abstract fun unionFlow(flows: Collection<FLOW>): FLOW
-
-    // -------------------------------- Flow mutators --------------------------------
+abstract class LogicSystem(protected val context: ConeInferenceContext) {
```
Ok. My reasoning for leaving them separate was that `PersistentLogicSystem` contains lots of rather large methods so `LogicSystem` makes the actual public interface more visible without having to query the IDE.

pyos commented on compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/CFGNode.kt:
```diff
@@ -654,7 +665,8 @@ class ConstExpressionNode(owner: ControlFlowGraph, override val fir: FirConstExp
 
-class CheckNotNullCallNode(owner: ControlFlowGraph, override val fir: FirCheckNotNullCall, level: Int, id: Int) : CFGNode<FirCheckNotNullCall>(owner, level, id) {
+class CheckNotNullCallNode(owner: ControlFlowGraph, override val fir: FirCheckNotNullCall, level: Int, id: Int)
+    : CFGNode<FirCheckNotNullCall>(owner, level, id), UnionNodeMarker {
```
Because it's implemented as a synthetic call so in theory it could be the first completed call and thus be a data flow sink for postponed lambdas inside `x`. I'm not sure if that's possible in practice.

demiurg906 replied:
Got it. Actually it might me something like that:
```kotlin
val x: Int = select(run { materialize() }, run { materialize() })!!
```

pyos commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt:
```diff
@@ -155,13 +151,10 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
     }
 
-    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirStatement> {
+    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirStatement>? {
```
This is actually a hint that the usages are somewhat strange - there's currently some code duplication and inconsistency in how return types of lambdas are inferred. @sfs was looking at this in the context of `@Composable` functions, I think. I tried to modify that code too, but ended up breaking several hacks in checkers so that's nowhere close to being ready.

pyos replied:
FYI, here's one example of how these inconsistencies manifest: https://youtrack.jetbrains.com/issue/KT-54742. Note how there is apparently code to handle this exact case, it just doesn't work for...reasons.

demiurg906 replied:
I didn't get how it's connected to the `returnExpressionsOfAnonymousFunction(orNull)`
Most users of `returnExpressionsOfAnonymousFunction` assumes that lambda already analyzed, so DFA knows all return expressions. And there is only one exception when this function can be called before/during lambda analysis. So I don't understand why exceptional behavior should be declared as default one in API)

pyos commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt:
```diff
@@ -1222,6 +1247,9 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
     private val CFGNode<*>.origin: CFGNode<*> get() = if (this is StubNode) firstPreviousNode else this
 
+    private val CFGNode<*>.livePreviousFlows: List<FLOW>
+        get() = if (isDead) previousNodes.map { it.flow } else previousNodes.mapNotNull { if (!it.isDead) it.flow else null }
```
I don't really like `runIf` because the name has the implication that it's for side effects, but ok.

pyos commented on compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/CFGNode.kt:
```diff
@@ -125,6 +126,12 @@ sealed class CFGNode<out E : FirElement>(val owner: ControlFlowGraph, val level:
+    open var flow: PersistentFlow
+        get() = _flow ?: throw IllegalStateException("flow for $this not initialized - traversing nodes in wrong order?")
+        @CfgInternals
+        set(value) { _flow = value } // TODO: forbid reassignment
```
FirDataFlowAnalyzer.kt is in a different module apparently.

demiurg906 replied:
Indeed

pyos commented on compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/PersistentLogicSystem.kt:
```diff
@@ -17,48 +17,21 @@ data class PersistentTypeStatement(
-        backwardsAliasMap = persistentMapOf()
-        assignmentIndex = persistentMapOf()
-    }
+class PersistentFlow internal constructor(
```
done

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt:
```diff
@@ -400,10 +402,10 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
-                            // flow.addImplication((expressionVariable eq !isType) implies (operandVariable notEq null))
-                        }
+                        } else if (type.isMarkedNullable) {
+                            // x !is T? => x != null
```
It's safe to keep

demiurg906 commented:
One module in `intellij-ultimate` stopped to compile. I'll investigate

pyos commented:
I'm not actually sure how to fix this. The data flow from `select { ... }` should clearly go into the exit of run's lambda, but the only reason it does that right now is that exit nodes are only collected in `exitAnonymousFunction` so `joinDataFlowFromPostponedLambdasWith` misses data flow from any lambda that has not been visited yet. So for example this compiles and breaks at runtime on master:

```
import kotlin.contracts.*

class Lateinit<R : Any> {
    lateinit var value: R
}

@OptIn(ExperimentalContracts::class)
public inline fun <R : Any> build(crossinline builder: Lateinit<R>.() -> Unit): R {
    contract { callsInPlace(builder, InvocationKind.EXACTLY_ONCE) }
    return Lateinit<R>().apply(builder).value
}

val p = false

fun main() {
  var y: Any? = 123
  val x: String = run {
    if (p)
      return@run build { y as String; value = "..." }
    else
      return@run ""
  } // call unifies data flow from lambda and two returns
  y.length
}
```

At this point I'm thinking that there needs to be a mechanism for "giving up", i.e. dropping data flow from called-in-place lambdas and telling `FirLocalVariableAssignmentAnalyzer` to assume the lambda has been leaked, in case things get too complicated to handle properly.

demiurg906 commented:
Can it be reproduced without builder inference? E.g. if explicitly specify type argument in `select`?

pyos commented:
I've added a commit to work around this issue for now. It's not a correct workaround, but it's no more incorrect that the existing code, and at least there's a test and a TODO now.

demiurg906 commented:
I think this workaround will be enough. I propose to not push too much into DFA and builder inference interaction, because we have plans to completely rework how BI is implemented somewhere after 1.9, so there might be a situation when current fixes will be useless with new implementation

demiurg906 commented:
Here are minimized cases which failed in intellij:
```kotlin
fun test_1() {
    var s: String? = null

    for (i in 1..10) {
        if (s == null) {
            s = "hello"
        }
        s.length // smartcast in K1, unsafe call in K2
        noInlineRun { s.length } // smartcast in K1, unsafe call in K2   <------------
    }
}

fun test_2_1() {
    var s: String? = null

    for (i in 1..10) {
        if (s == null) {
            s = "hello"
        }
        s.length // smartcast in K1, unsafe call in K2
        noInlineRun { s.length } // unsafe call in K1 and K2
        s = ""
    }
}

fun test_2_2() {
    var s: String? = null

    for (i in 1..10) {
        if (s == null) {
            s = "hello"
        }
        s.length // smartcast in K1, unsafe call in K2
        noInlineRun { s.length } // unsafe call in K1 and K2
        s = null
    }
}

fun test_3_1() {
    var s: String? = null

    for (i in 1..10) {
        if (s == null) {
            s = "hello"
        } else {
            s = null
        }
        s.length // unsafe call in K1 and K2
        noInlineRun { s.length } // unsafe call in K1 and K2
    }
}

fun test_3_2() {
    var s: String? = null

    for (i in 1..10) {
        if (s == null) {
            s = "hello"
        } else {
            s = "world"
        }
        s.length // smartcast in K1 and K2
        noInlineRun { s.length } // smartcast in K1, unsafe call in K2   <------------
    }
}

fun noInlineRun(block: () -> Unit) {}
```

```kotlin
fun test() {
    for (x in 1..10) { // removal of loop fixes the issue
        var s: String? = getNullableString()
        if (s!!.endsWith("hello")) {
            s = if (getBoolean()) {
                getString()
            } else {
                getNullableString()
            }
            if (s != null) {
                noInlineRun {
                    s.length // smarcast in K1, smartcast impossible in K2
                }
            }
        }
    }
}

fun getNullableString(): String? = null
fun getString(): String = ""
fun getBoolean(): Boolean = true
fun noInlineRun(block: () -> Unit) {}
```

Could you please take a look on them?

pyos commented:
The last one is a K1 bug, it should not compile: https://pl.kotl.in/phs-PV0zf

pyos commented:
Hmmm, this is somewhat hard. K2 is actually more correct in the first example... https://pl.kotl.in/8mJhfLCYM https://pl.kotl.in/51hznnxXU

The last one I think I can fix (filter out `s` from variables at loop start since it's declared inside the loop).

pyos commented:
Fixed the last case. Also added a test that covers the rest. It's probably better to patch intellij - while theoretically FIR can prove that the smart cast is valid, it requires basically checking that it's valid at any point after the first assignment in the loop.

demiurg906 commented:
Thank you!

I've created [KT-55338](https://youtrack.jetbrains.com/issue/KT-55338) for the second case

demiurg906 commented:
After several rebases and reruns of CI I finally merged your PR. Thank you once again

demiurg906 commented:
May I ask you to take a look at the problems with CFG building for local declarations ([KT-48746](https://youtrack.jetbrains.com/issue/KT-48746) if it fits with your plans?

demiurg906 commented:
There is another interesting case:
```kotlin
fun test() {
    var x: String? = null
    object {
        init {
            x = "hello"
        }
    }
    x.length // theoretically we can have smartcast here
    
    x = null
    
    class Local {
        init {
            x = "hello"
        }
    }
    x.length // but not here
}
```
